openff/toolkit/typing/engines/smirnoff/parameters.py

```diff
@@ -5839,6 +6331,74 @@ def _create_openmm_virtual_particle(self, system, force, top_mol, vsite, ref_key
         return ids
 
 
+class _ParameterTermKey(tuple):
+    """
+    A tuple typedef of nested tuples that aims to give a serialized key to uniquely
+    identify every term in the FF.
+
+    This tuple therefore uniquely identifies every individual term in either the
+    FF/ParameterHandler/ParameterType. This structure gives a quasi-regular way
+    to access any of these terms. Processing/handling of these keys are slightly
+    different for each object, and therefore the shape of this key will depend
+    on who/what generated it. For example, term keys produced by FF objects
+    should only be handled by FF objects, and likewise for subsequent levels.
+
+    The only normal structure to this tuple is that it is recursively defined as
+
+    T[x] = (str, T[x-1])
+    T[0] = (str,)
```

Comment:
What are they types in these tuples? Is every tuple guaranteed to contain only another tuple or string down to the end, in which it's a `ParameterTerm`?

Reply:
Yes that's right, I only ever put `str` in them, and the tuples always length 2, except for the last case. It's not guaranteed, per se, since it is essentially a struct/dataclass with no validators, but down the road if this is solidified we can treat it more like a real encapsulating object with testable behavior.

Reply:
👍 I thought briefly about the type annotation that would describe that, but quickly realized it's either not possible, so complex to be useless, or just too much for me to wrap my head around.